src/python/pants/engine/isolated_process.py

```diff
@@ -86,11 +87,61 @@ def _verify_env_is_dict(cls, env):
 
 
 class ExecuteProcessResult(datatype(['stdout', 'stderr', 'exit_code', 'output_directory_digest'])):
-  pass
+  """Result of executing a process. Will raise an exception if the exit code is non-zero."""
+
+
+class FallibleExecuteProcessResult(datatype(['stdout', 'stderr', 'exit_code', 'output_directory_digest'])):
+  """An ExecuteProcessResult that will not raise an exception if the exit code is non-zero."""
+
+
+class ProcessExecutionFailure(Exception):
+  """Used to denote that a process exited, but was unsuccessful in some way.
+
+  For example, exiting with a non-zero code.
+  """
+
+  MSG_FMT = """process '{desc}' failed with code {code}.
+stdout:
+{stdout}
+stderr:
+{stderr}
+"""
+
+  def __init__(self, exit_code, stdout, stderr, process_description):
+    # These are intentionally "public" members.
+    self.exit_code = exit_code
+    self.stdout = stdout
+    self.stderr = stderr
+
+    msg = self.MSG_FMT.format(
+      desc=process_description, code=exit_code, stdout=stdout, stderr=stderr)
+
+    super(ProcessExecutionFailure, self).__init__(msg)
+
+
+@rule(ExecuteProcessResult, [Select(FallibleExecuteProcessResult), Select(ExecuteProcessRequest)])
+def fallible_to_exec_result_or_raise(fallible_result, request):
+  """Converts a FallibleExecuteProcessResult to a ExecuteProcessResult or raises an error."""
+
+  if fallible_result.exit_code == 0:
+    return ExecuteProcessResult(
```

Comment:
how would you feel about plumbing this `raise on exit_code != 0` behavior as a defaulted parameter in an `ExecuteProcessRequest` with some override capability?

I think there will definitely be cases where we'll hit non-zero exit codes in a process execution but want to continue running.

one example would be: aggregating results of hundreds of pytest executions without e.g. `--fast-fail`. rather than Throw'ing on the first failure, we may want to aggregate those in a `@console_rule` and then fail at that level.

Reply:
Hm. I was going off of the design discussion in #5719.
 
You either request `ExecuteProcessRequest`, in which case you get an exception, or you request `FallibleExecuteProcessRequest` and you get an object you can inspect for failures.

In cases where we want to inspect the exit_code, we'd just request the `FallibleExecuteProcessRequest` one

Reply:
yep, my bad - realized that just after I posted.